File: Sprint4/L/C++/main.cpp
// https://contest.yandex.ru/contest/23991/problems/L/
#include <iostream>
#include <string>
#include <unordered_map>
#include <vector>
#include <cstdint>

const uint64_t base = 31, mod = 0xFFFFFFFFFFFFFFFF;

uint64_t get_hash(const std::vector<uint64_t> &hashes,
                   uint64_t powers,
                   int i, int n) {
    uint64_t hash = hashes[i + n] - hashes[i] * powers % mod;
    hash = (hash % mod + mod) % mod;
    return hash;
}

int main() {
    int n, k;
    std::string s;
    std::cin >> n >> k;
    std::cin >> s;

    std::unordered_map<uint64_t, int> map;
    std::unordered_map<uint64_t, int> result;

    std::vector<uint64_t> hashes(s.size() + 1, 0);
    uint64_t powers = 1;
    for (int i = 1; i <= s.size(); ++i) {
        hashes[i] = (hashes[i - 1] * base + s[i - 1] - 'a' + 1) % mod;
        if (i <= n) {
            powers = (powers * base) % mod;
        }
    }

    for (int i = 0; i + n <= s.size(); i++) {
        auto hash = get_hash(hashes, powers, i, n);
        if (++map[hash] == 1) {
            result[hash] = i;
        }
    }

    for (auto &pair : map) {
        if (pair.second >= k) {
            std::cout << result[pair.first] << " ";
        }
    }

    return 0;
}

File: Sprint4/L/Rust/src/main.rs
// https://contest.yandex.ru/contest/23991/problems/L/
use std::collections::HashMap;
use std::io;

const BASE: u64 = 31;
const MOD: u64 = u64::MAX;

fn get_hash(hashes: &[u64], powers: u64, i: usize, n: usize) -> u64 {
    let mut hash = hashes[i + n].wrapping_sub(hashes[i].wrapping_mul(powers).wrapping_rem(MOD));
    hash = (hash.wrapping_rem(MOD).wrapping_add(MOD)).wrapping_rem(MOD);
    hash
}

fn main() {
    let mut input = String::new();
    io::stdin().read_line(&mut input).unwrap();
    let mut nums = input.trim().split_whitespace();
    let n: usize = nums.next().unwrap().parse().unwrap();
    let k: usize = nums.next().unwrap().parse().unwrap();

    input.clear();
    io::stdin().read_line(&mut input).unwrap();
    let s = input.trim();

    let mut map = HashMap::new();
    let mut result = HashMap::new();
    let mut hashes = vec![0u64; s.len() + 1];
    let mut powers = 1u64;
    for i in 1..=s.len() {
        hashes[i] = (hashes[i - 1].wrapping_mul(BASE) + s.chars().nth(i - 1).unwrap() as u64 - 'a' as u64 + 1).wrapping_rem(MOD);
        if i <= n {
            powers = (powers.wrapping_mul(BASE)).wrapping_rem(MOD);
        }
    }

    for i in 0..=s.len() - n {
        let hash = get_hash(&hashes, powers, i, n);
        let counter = map.entry(hash).or_insert(0);
        *counter += 1;
        if *counter == 1 {
            result.insert(hash, i);
        }
    }

    for (hash, count) in map.iter() {
        if *count >= k {
            print!("{} ", result[hash]);
        }
    }
}

